modules/indexing/src/test/java/org/apache/ignite/internal/processors/query/h2/IgniteSqlQueryStartFinishListenerTest.java

```diff
@@ -0,0 +1,495 @@
+
+            GridQueryStartedInfo startedInfo = qryStarted.get();
+            assertEquals(sysSchemaName(), startedInfo.schemaName());
+            assertEquals(qry, startedInfo.query());
+            assertEquals(grid(SERVER_NODE_NAME).localNode().id(), startedInfo.nodeId());
+            assertEquals(false, startedInfo.local());
+            assertEquals(GridCacheQueryType.SQL_FIELDS, startedInfo.queryType());
+
+            assertWithTimeout(qryFinished::get, is(notNullValue()), 1_000);
+
+            GridQueryFinishedInfo finishedInfo = qryFinished.get();
+            assertEquals(sysSchemaName(), finishedInfo.schemaName());
+            assertEquals(qry, finishedInfo.query());
+            assertEquals(grid(SERVER_NODE_NAME).localNode().id(), startedInfo.nodeId());
+            assertEquals(false, finishedInfo.local());
+            assertEquals(GridCacheQueryType.SQL_FIELDS, finishedInfo.queryType());
+            assertEquals(false, finishedInfo.failed());
+            assertThat(finishedInfo.finishTime() - finishedInfo.startTime(), is(greaterOrEqualTo(delay)));
+
+            qryStarted.set(null);
+            qryFinished.set(null);
+        }
+
+        {
+            final String schema = "TEST1";
+            final String qry = "select \"default\".can_fail() from " + sysSchemaName() + ".caches where ? = ? limit 1";
+
+            GridTestUtils.SqlTestFunctions.fail = true;
+
+            GridTestUtils.assertThrowsWithCause(() -> execSqlLocal(schema, qry, 1, 1), IgniteSQLException.class);
+
+            assertWithTimeout(qryStarted::get, is(notNullValue()), 1_000);
+
+            GridQueryStartedInfo startedInfo = qryStarted.get();
+            assertEquals(schema, startedInfo.schemaName());
+            assertEquals(qry, startedInfo.query());
+            assertEquals(grid(SERVER_NODE_NAME).localNode().id(), startedInfo.nodeId());
+            assertEquals(true, startedInfo.local());
+            assertEquals(GridCacheQueryType.SQL_FIELDS, startedInfo.queryType());
+
+            assertWithTimeout(qryFinished::get, is(notNullValue()), 1_000);
+
+            GridQueryFinishedInfo finishedInfo = qryFinished.get();
+            assertEquals(schema, finishedInfo.schemaName());
+            assertEquals(qry, finishedInfo.query());
+            assertEquals(grid(SERVER_NODE_NAME).localNode().id(), startedInfo.nodeId());
+            assertEquals(true, finishedInfo.local());
+            assertEquals(GridCacheQueryType.SQL_FIELDS, finishedInfo.queryType());
+            assertEquals(true, finishedInfo.failed());
+            assertThat(finishedInfo.finishTime(), is(greaterOrEqualTo(finishedInfo.startTime())));
+
+            qryStarted.set(null);
+            qryFinished.set(null);
+        }
+
+        {
+            final String qry = "text query";
+
+            IgniteCache<String, Object> cache = grid(CLIENT_NODE_NAME).cache(DEFAULT_CACHE_NAME);
+
+            cache.query(new TextQuery<String, String>(String.class, "text query"));
```

Comment:
I don't quite understand, if we issue `TextQuery` over cache got from client node, why do we later check that `startedInfo.nodeId()` equals to server node id?

Reply:
it is because nodeId represents an id of the node which starts execute the query, not an id of the node which sends request. Semantic is the same as for running query info (for running queries view)